pallet/account-migration/src/lib.rs

```diff
@@ -16,174 +16,167 @@
-	use frame_system::pallet_prelude::*;
 
 	#[pallet::pallet]
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::config]
-	pub trait Config: frame_system::Config + pallet_evm::Config {
-		/// The overarching event type.
+	pub trait Config:
+		frame_system::Config<
+		AccountId = AccountId20,
+		Index = Index,
+		AccountData = AccountData<Balance>,
+	>
+	{
+		/// Override the [`frame_system::Config::RuntimeEvent`].
 		type RuntimeEvent: From<Event> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
-		/// Currency type for the runtime.
-		type Currency: Currency<Self::AccountId>;
-	}
-
-	// Store the migrated balance snapshot for the darwinia-1.0 chain state.
-	#[pallet::storage]
-	#[pallet::getter(fn balance_of)]
-	pub(super) type Balances<T> = StorageMap<_, Blake2_128Concat, AccountId32, Balance>;
-
-	#[pallet::error]
-	pub enum Error<T> {
-		/// This account does not exist in the darwinia 1.0 chain state.
-		AccountNotExist,
 	}
 
+	#[allow(missing_docs)]
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(super) fn deposit_event)]
 	pub enum Event {
-		/// Claim to the new account id.
-		Claim { old_pub_key: AccountId32, new_pub_key: H160, amount: Balance },
+		/// An account has been migrated.
+		Migrated { from: AccountId32, to: AccountId20 },
 	}
 
-	#[pallet::genesis_config]
-	#[cfg_attr(feature = "std", derive(Default))]
-	pub struct GenesisConfig {
-		pub migrated_accounts: Vec<(AccountId32, Balance)>,
-	}
+	/// [`frame_system::Account`] data.
+	#[pallet::storage]
+	#[pallet::getter(fn account_of)]
+	pub type Accounts<T: Config> =
+		StorageMap<_, Identity, AccountId32, AccountInfo<Index, AccountData<Balance>>>;
 
-	#[pallet::genesis_build]
-	impl<T: Config> GenesisBuild<T> for GenesisConfig {
-		fn build(&self) {
-			self.migrated_accounts.iter().for_each(|(account, amount)| {
-				Balances::<T>::insert(account, amount);
-			});
-		}
-	}
+	// TODO: identity storages
+	// TODO: proxy storages
+	// TODO: staking storages
+	// TODO: vesting storages
 
 	#[pallet::call]
-	impl<T: Config> Pallet<T>
-	where
-		AccountIdOf<T>: From<H160>,
-		BalanceOf<T>: From<Balance>,
-	{
-		// since signature and chain_id verification is done in `validate_unsigned`
-		// we can skip doing it here again.
-		// TODO: update weight
+	impl<T: Config> Pallet<T> {
+		/// Migrate all the account data under the `from` to `to`.
 		#[pallet::weight(0)]
-		pub fn claim_to(
+		pub fn migrate(
 			origin: OriginFor<T>,
-			_chain_id: u64,
-			old_pub_key: AccountId32,
-			new_pub_key: H160,
-			_sig: Signature,
+			from: AccountId32,
+			to: AccountId20,
+			_signature: Signature,
 		) -> DispatchResult {
 			ensure_none(origin)?;
 
-			let Some(amount) = Balances::<T>::take(&old_pub_key) else  {
-				return Err(Error::<T>::AccountNotExist.into());
-			};
+			let account = <Accounts<T>>::take(from)
+				.ok_or("[pallet::account-migration] already checked in `pre_dispatch`; qed")?;
+
+			<frame_system::Account<T>>::insert(to, account);
```

Comment:
Does the account taken out here come from the darwinia 1.0?  As we discussed in another place, the `to` refers to a fresh account. Then you overwrite existing account info to a fresh account, this is very tricky.  

Image this:

1. A user wants to claim a frequently used account in darwinia 1.0, he creates a fresh H160 account.
2. Send an unsigned migrate extrinsic, then his H160 account balance is the same as the old one, but the nonce is also the same as the old one, starting from a strange number.  This looks weird.

Reply:
> he creates a fresh H160 account

This should be an address genereated from some key, instead of onchain account, so the account is not existings yet.

As a result, new account can be created with non-zero nonce from migrations. Will that cause any security issue? Can not foresee any. @hujw77 Can you help review this?

Reply:
In this case, the account nonce should be zero.

Reply:
> In this case, the account nonce should be zero.

Any issue if non zero?

Reply:
> But my questions is, if some evm address in Darwinia 1.0 receive some funds, but did not initialiated any tx, will the system create an account for it?

It will

Reply:
I think it is safety, but it may break nonce's meaning. 
`eth_getTransactionCount` returns the number of transactions sent from an address.

Reply:
As a mention, new account from nonce > 0, break the assuming that deployer want to deploy same contract address on multi-chain.